### ISSUE TITLE
KeyPair.fromAccountId throws  java.lang.IllegalArgumentException: not a valid GroupElement

### ISSUE THREAD

**Issue by efritze** [stellar/java-stellar-sdk#229]
`
KeyPair.fromAccountId("GBF5EIG7LVPCFFU5PYNIF47GH6CQRTOGTSX3KRBZ5MZBMOI5TAUNN4MT");
`

The line above doesn't throw any exceptions.

`
KeyPair.fromAccountId("GAH6H2XPCZS27WMKPTZJPTDN7JMBCDHTLU5WQP7TUI2ORA2M5FY5DHNU");
`

The line above throws the exception below:

`
        java.lang.IllegalArgumentException: not a valid GroupElement
	at shadow.net.i2p.crypto.eddsa.math.GroupElement.<init>(GroupElement.java:336)
	at shadow.net.i2p.crypto.eddsa.math.GroupElement.<init>(GroupElement.java:287)
	at shadow.net.i2p.crypto.eddsa.spec.EdDSAPublicKeySpec.<init>(EdDSAPublicKeySpec.java:36)
	at org.stellar.sdk.KeyPair.fromPublicKey(KeyPair.java:115)
	at org.stellar.sdk.KeyPair.fromAccountId(KeyPair.java:106)
`

I'd say it's a pretty urgent issue.

**Comment by bartekn** [comment]
Duplicate of #131. Some byte arrays are not valid ed25519 public keys.

**Comment by ire-and-curses** [comment]
@bartekn I looked through the linked issue. What's the next step with this? `GAH6H2XPCZS27WMKPTZJPTDN7JMBCDHTLU5WQP7TUI2ORA2M5FY5DHNU` is a valid address - it's the friendbot account.

**Comment by bartekn** [comment]
@ire-and-curses are you sure? This is the current friendbot account: `GAIH3ULLFQ4DGSECF2AR555KZ4KNDGEKN4AFI4SU2M7B43MGK3QJZNSR`. Anyway @efritze opened #230 and that one is valid. Stellar-Core doesn't check if a newly created account has a valid pubkey. So it's possible that accounts in the ledger have invalid pubkeys. We need to remove usage of `Keypair` when parsing Horizon responses as they may contain such pubkeys.

**Comment by ire-and-curses** [comment]
Oh, you're right. It's [this address](https://stellar.expert/explorer/public/account/GAH6H2XPCZS27WMKPTZJPTDN7JMBCDHTLU5WQP7TUI2ORA2M5FY5DHNU).

That's amazing that there can be invalid public keys for created accounts. Other SDKs are probably affected also.

It seems like stellar-core should check for validity?! Is there an issue for that?

**Comment by tomquisel** [comment]
The conclusion here is that it's not a good assumption that `accountId`s will necessarily be valid public keys. Clients should be looking at `KeyPair`s constructed from signers for an account, instead of directly from the `accountId`. Deterministic accounts and other future improvements will break that assumption.

So the fix is to avoid calling `KeyPair` on accountIds in clients and SDKs. Does that sound right @bartekn ?

**Comment by tamirms** [comment]
> So the fix is to avoid calling KeyPair on accountIds in clients and SDKs

@tomquisel @bartekn it looks like that approach was taken here https://github.com/stellar/java-stellar-sdk/pull/139

but that PR was later reverted. @bartekn do you remember why it was necessary to revert the PR?

**Comment by bartekn** [comment]
@tomquisel this is correct.

@tamirms good find. I remember I was working on this but I didn't have time in the last few days to check the final status of this issue (I was thinking that maybe I forgot to remove some instances of `KeyPair` in the last PR). Now, I know what happened. At that time we didn't have any branch strategy in java-sdk. A short time after #139 was merged, we had to push an important fix (bug in shadowing?) and release it in a patch release. However, #139 was a breaking change not suitable for a patch release so I reverted it.
